src/doc/trpl/the-stack-and-the-heap.md

```diff
@@ -1,3 +1,526 @@
+| 1       | i    | 2 ^ 30 |
+| 0       | h    | 3      |
+
+We’ve allocated memory for `f` and `g`. `baz()` is very short, so when it’s
+over, we get rid of its stack frame:
+
+| Address | Name | Value  |
++---------+------+--------+
+| 2 ^ 30  |      | 20     |
+| ...     | ...  | ...    |
+| 5       | z    | 4      |
+| 4       | y    | 10     |
+| 3       | x    | 0      |
+| 2       | j    | 0      |
+| 1       | i    | 2 ^ 30 |
+| 0       | h    | 3      |
+
+Next, `foo()` calls `bar()` with `x` and `z`:
+
+| Address      | Name | Value        |
++--------------+------+--------------+
+|  2 ^ 30      |      | 20           |
+| (2 ^ 30) - 1 |      | 5            |
+| ...          | ...  | ...          |
+| 10           | e    | 4            |
+| 9            | d    | (2 ^ 30) - 1 |
+| 8            | c    | 5            |
+| 7            | b    | 4            |
+| 6            | a    | 0            |
+| 5            | z    | 4            |
+| 4            | y    | 10           |
+| 3            | x    | 0            |
+| 2            | j    | 0            |
+| 1            | i    | 2 ^ 30       |
+| 0            | h    | 3            |
+
+We end up allocating another value on the heap, and so we have to subtract one
+from `2 ^ 30`. It’s easier to just write that than `1,073,741,823`. In any
+case, we set up the variables as usual.
+
+We got up to address `10`! Usually, memory addresses are referred to with a
+hexadecimal number, rather than a decimal one. Here’s a conversion to
+hexadecimal:
+
+| Address    | Name | Value      |
++------------+------+------------+
+| 0x40000000 |      | 20         |
+| 0x3FFFFFFF |      | 5          |
+| ...        | ...  | ...        |
+| 0x0000000a | e    | 4          |
+| 0x00000009 | d    | 0x3FFFFFFF |
+| 0x00000008 | c    | 5          |
+| 0x00000007 | b    | 4          |
+| 0x00000006 | a    | 0          |
+| 0x00000005 | z    | 4          |
+| 0x00000004 | y    | 10         |
+| 0x00000003 | x    | 0          |
+| 0x00000002 | j    | 0          |
+| 0x00000001 | i    | 0x40000000 |
+| 0x00000000 | h    | 3          |
+
+This is nicer in some ways and worse in others. It’s a bit more terse for the
+big numbers, but more verbose for the smaller ones. We can get rid of extra
+zeroes:
+
+| Address    | Name | Value      |
++------------+------+------------+
+| 0x40000000 |      | 20         |
+| 0x3FFFFFFF |      | 5          |
+| ...        | ...  | ...        |
+| 0x0a       | e    | 4          |
+| 0x09       | d    | 0x3FFFFFFF |
+| 0x08       | c    | 5          |
+| 0x07       | b    | 4          |
+| 0x06       | a    | 0          |
+| 0x05       | z    | 4          |
+| 0x04       | y    | 10         |
+| 0x03       | x    | 0          |
+| 0x02       | j    | 0          |
+| 0x01       | i    | 0x40000000 |
+| 0x00       | h    | 3          |
+
+Regardless, you may see diagrams in other examples that use hexadecimal. We’ll
+just stick to decimal, since we’re already halfway through the example.
```

Comment:
I think that it is a mistake to mix using the abstract word-oriented addresses with hexadecimal notation -- i.e., if you're going to talk about how "memory addresses are usually referred to", then I would also include the fact that they typically are byte-addresses, while this presentation has been using word-addressed memory.

I would normally argue for switching to hexadecimal from the beginning... the only reason I can see not to shift to hexadecimal for addresses from the beginning is because it may be harder to read, especially if you write out all the zeroes (so that the least-significant digits line up).

(One *could* address that by using underscore instead of zero for the leading zero digits, like so:
`0x______10, 0x______14, 0x______18`, ...; note that I also here switched to byte-oriented addresses.)

But in the end I'm happy leaving this up to @steveklabnik